src/App.tsx

```diff
@@ -1,61 +1,218 @@
+      if (user.id === todo.userId) {
+        todosAcc.push({ ...todo, ...{ createdAt: new Date() } });
+
+        return todosAcc;
+      }
+
+      return todosAcc;
+    }, []);
+
+    return {
+      userId: user.id,
+      userName: user.name,
+      userEmail: user.email,
+      userLogin: user.username,
+      todoPosts: userTodo,
+    };
+  })
+  .filter(userTodo => userTodo.todoPosts.length > 0);
+
+function findTheBiggestPostId(postList: UserTodos[]): number {
+  return Math.max(
+    ...postList.map(userPost =>
+      Math.max(...userPost.todoPosts.map(post => post.id)),
+    ),
+  );
+}
 
 export const App = () => {
+  // #region baseStates
+  const [currentUsersTodos, setCurrentUsersTodos] =
+    useState<UserTodos[]>(usersTodos);
+  const [postCount, setPostCount] = useState(
+    findTheBiggestPostId(currentUsersTodos) + 1,
+  );
+  // #endregion
+  // #region title
+  const [title, setTitle] = useState('');
+  // #endregion
+  // #region userSelecting
+  const [isUserSelected] = useState(false);
+  const [userSelectingValue, setUserSelectingValue] = useState<number>(0);
+  // #endregion
+
+  // #region fieldErrors
+  const [hasTitleAnError, setHasTitleAnError] = useState<boolean>(false);
+  const [hasSelectedUserAnError, setHasSelectedUserAnError] =
+    useState<boolean>(false);
+
+  const titleFieldHasAnError = title.length === 0;
+  const selectedUserFieldHasAnError = userSelectingValue === 0;
+  const formHasSomeError = hasTitleAnError || hasSelectedUserAnError;
+  // #endregion
+
+  // #region handlers
+  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+    setHasTitleAnError(false);
+  };
+
+  const handleSelectingUser = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setUserSelectingValue(+event.target.value);
+    setHasSelectedUserAnError(false);
+  };
+
+  const handleSelectedUser = (userId: number): User => {
+    return usersFromServer.find(user => user.id === userId) as User;
+  };
+
+  const onSubmit = (selectedUser: User): void => {
+    const existingUser = currentUsersTodos.find(
+      user => user.userId === selectedUser.id,
+    );
+
+    const newTodo = {
+      id: postCount,
+      title: title,
+      completed: false,
+      userId: selectedUser.id,
+      createdAt: new Date(),
+    };
+
+    if (existingUser) {
+      existingUser.todoPosts.push(newTodo);
+
+      return setCurrentUsersTodos([...currentUsersTodos]);
+    }
+
+    return setCurrentUsersTodos([
+      ...currentUsersTodos,
+      {
+        userId: selectedUser.id,
+        userName: selectedUser.name,
+        userEmail: selectedUser.email,
+        userLogin: selectedUser.username,
+        todoPosts: [
+          {
+            id: postCount,
+            title: title,
+            completed: false,
+            userId: selectedUser.id,
+            createdAt: new Date(),
+          },
+        ],
+      },
```

Comment:
When you need to update the state depending on the previous one use functional `setState`, and you don't need to store this data in the state at all

Reply:
Here If I don't have definite user with any post, I update `currentUsersTodos` and add new user (selected user) with new post in his post's history